File: Python/面试经典150题/链表/2_两数相加.py
from typing import Optional
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)
        current = dummy
        up = 0

        while l1 or l2 or up:
            a = l1.val if l1 else 0
            b = l2.val if l2 else 0

            total = a + b + up
            current.next = ListNode(total % 10)
            up = total // 10

            if l1:
                l1 = l1.next
            if l2:
                l2 = l2.next
            current = current.next

        return dummy.next

File: JS/LeetCode热题HOT100/2_两数相加.js
function ListNode(val, next) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : next)
}

var addTwoNumbers = function(l1, l2) {
    const l3 = new ListNode(0)
    let p1 = l1, p2 = l2, p3 = l3
    let sum = 0, t = 0
    while(p1 || p2 || t){
        const v1 = p1 ? p1.val : 0
        const v2 = p2 ? p2.val : 0
        sum = v1 + v2 + t
        t = Math.floor(sum / 10)
        p3.next = new ListNode(sum%10)
        if(p1) p1 = p1.next
        if(p2) p2 = p2.next
        p3 = p3.next
    }
    return l3.next
};